src/base/orderedset.h

```diff
@@ -0,0 +1,121 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2021  Mike Tzou (Chocobo1)
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <functional>
+#include <set>
+
+#include "algorithm.h"
+
+template <typename T, typename Compare = std::less<T>>
+class OrderedSet : public std::set<T, Compare>
+{
+    using ThisType = OrderedSet<T, Compare>;
+
+public:
+    using BaseType = std::set<T, Compare>;
+
+    using key_type = typename BaseType::key_type;
+    using value_type = typename BaseType::value_type;
+
+    using BaseType::BaseType;
+    using BaseType::operator=;
+
+    // The following are custom functions that are in line with Qt API interface, such as `QSet`
+
+#if __cplusplus < 202002L
+    bool contains(const key_type &value) const
+    {
+        return (BaseType::find(value) != BaseType::cend());
+    }
+#endif
+
+    int count() const
+    {
+        return static_cast<int>(BaseType::size());
+    }
+
+    ThisType &intersect(const ThisType &other)
+    {
+        Algorithm::removeIf(*this, [&other](const T &value) -> bool
+        {
+            return !other.contains(value);
+        });
+        return *this;
+    }
+
+    bool isEmpty() const
+    {
+        return BaseType::empty();
+    }
+
+    template <typename Container, typename Separator>
+    Container join(const Separator &separator) const
```

Comment:
Sorry, it is unclear what Container it should be. In Qt, it is common to use this term in relation to array/list/map types etc. It's really supposed to transform one container (set) to some another?

Reply:
>It's really supposed to transform one container (set) to some another?

`QStringList::join()` is converting a list type to another (`QString`) type. So we may view `join()` operation as some kind of a transform operation (in the sense of `std::transform`).
And `OrderedSet<T>` takes a type parameter so it depends on the caller to define what should be the result (`Container`). IMO it is too restricted to always assume the return type is `T`.

Reply:
>`QStringList::join()` is converting a list type to another (`QString`) type. So we may view `join()` operation as some kind of a transform operation (in the sense of `std::transform`).

This doesn't look like a valid analogy to me.
`std::transform` transforms `Container<A>` of size `N` to `Container<B>` of the same size, isn't it?
`join` is not a general algorithm for different containers, but only for string-like ones (no wonder it is not defined in `QList`, but only in `QStringList`). It looks more like a generalization of string concatenation.
The algorithm you're suggesting here certainly makes sense on its own, but it definitely shouldn't be called a `join`. It has meaning of `join` only in some special cases.

Reply:
>IMO it is too restricted to always assume the return type is `T`.

I would say that this is most likely exactly what everyone expects (or used to expect) from `join()`.
Join: `Array<T>` => `T`
Split: `T` => `Array<T>` (reverse operation to Join)
(`Array` above is any array like container.)

Reply:
>std::transform transforms Container<A\> of size N to Container<B\> of the same size, isn't it?

If there are no `std::back_inserter` involved, otherwise the element count would be different.

>join is not a general algorithm for different containers, but only for string-like ones 

Then this function would need to be move to a specialization of `OrderedSet<QString, ...>` otherwise `join()` on generalized `T` doesn't make sense, isn't?
I think I picked the name `Container` poorly, in my own definition `QString` and `QByteArray` could also be viewed as some sort of "container".

>The algorithm you're suggesting here certainly makes sense on its own, but it definitely shouldn't be called a join. It has meaning of join only in some special cases.

I had the idea of turning this version of `join()` into a non-class template function that could adapt to a wide range of "containers" but I wasn't sure enough.
IMO this function resemble to me as a `transform` + `join` joint operation, but it is hard to find a appropriate name for it.

Reply:
Or I could change it to `T intersperse(const T &separator) const`
The function name comes from https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-List.html#v:intersperse

Reply:
I would prefer to see in the code the regular `join` in its regular string-related meaning. I think it'll be more readable and maintainable in that way.

Reply:
OK, I updated.